test/tokenizedSubaccounts/PPTSA/PPTSA_Validation.t.sol

```diff
@@ -0,0 +1,502 @@
+    tsa.signActionData(action1, "");
+
+    assertEq(tsa.lastSeenHash(), tsa.getActionTypedDataHash(action1));
+
+    IActionVerifier.Action memory action2 = _createDepositAction(2e18);
+
+    vm.prank(signer);
+    tsa.signActionData(action2, "");
+
+    assertEq(tsa.lastSeenHash(), tsa.getActionTypedDataHash(action2));
+
+    vm.expectRevert(IActionVerifier.OV_InvalidSignature.selector);
+    _submitToMatching(action1);
+
+    // TODO: Can withdraw even with a pending deposit action. Can lead to pending deposits being moved to TSA...
+    tsa.requestWithdrawal(10e18);
+    tsa.processWithdrawalRequests(1);
+
+    // Fails as no funds were actually deposited, but passes signature validation
+    vm.expectRevert("ERC20: transfer amount exceeds balance");
+    _submitToMatching(action2);
+  }
+
+  // Todo: duplicate of CCTSA validation. Possible merge?
+  function testPPTInvalidModules() public {
+    _depositToTSA(1e18);
+
+    vm.startPrank(signer);
+
+    IActionVerifier.Action memory action = _createDepositAction(1e18);
+    action.module = IMatchingModule(address(10));
+
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidModule.selector);
+    tsa.signActionData(action, "");
+
+    action.module = depositModule;
+    tsa.signActionData(action, "");
+    vm.stopPrank();
+  }
+
+  /////////////////
+  // Withdrawals //
+  /////////////////
+  function testPPTWithdrawalBaseAssetValidation() public {
+    _depositToTSA(3e18);
+    vm.startPrank(signer);
+
+    // correctly verifies withdrawal actions.
+    IActionVerifier.Action memory action = _createWithdrawalAction(3e18);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidBaseBalance.selector);
+    tsa.signActionData(action, "");
+
+    // reverts for invalid assets.
+    action.data = _encodeWithdrawData(3e18, address(11111));
+    vm.expectRevert(PrincipalProtectedTSA.PPT_InvalidAsset.selector);
+    tsa.signActionData(action, "");
+
+    vm.stopPrank();
+
+    // add a trade
+    uint expiry = block.timestamp + 1 weeks;
+    _executeDeposit(3e18);
+    _tradeRfqAsTaker(1e18, 1e18, expiry, 2000e18, 4.0e18, 1600e18, true);
+
+    action = _createWithdrawalAction(3e18);
+    vm.prank(signer);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_WithdrawingWithOpenTrades.selector);
+    tsa.signActionData(action, "");
+
+    vm.warp(block.timestamp + 8 days);
+    _setSettlementPrice("weth", uint64(expiry), 1500e18);
+    srm.settleOptions(markets["weth"].option, tsa.subAccount());
+
+    vm.startPrank(signer);
+    // now try to withdraw all of the base asset. Should fail
+    action = _createWithdrawalAction(3e18);
+    vm.expectRevert(PrincipalProtectedTSA.PPT_WithdrawingUtilisedCollateral.selector);
+    tsa.signActionData(action, "");
+
+    // now try to a small 5% of the base asset. Should pass
+    action = _createWithdrawalAction(0.15e18);
+    tsa.signActionData(action, "");
+
+    vm.stopPrank();
+  }
+
+  /////////////////
+  //    Trades   //
+  /////////////////
```

Comment:
nitpick: but much easier on reviewer and for yourself if you split out all these subsections into separate test files. Otherwise reviewer needs to scroll throw gigantic file. 

I know you follow CCTSA tests, but it's simpler to have subfolders:

- `admin.t.sol`
- `base.t.sol`
- `collateralTrade.t.sol`
- `rfq.t.sol`
- `deposit.t.sol`
- `withdrawal.t.sol`

Reply:
I agree with this. Just was trying to follow OG conventions.